packages/server/lib/server-base.ts

```diff
@@ -192,11 +221,90 @@ export abstract class ServerBase<TSocket extends SocketE2E | SocketCt> {
     this.socket.localBus.on('request:sent:with:credentials', this.resourceTypeAndCredentialManager.set)
   }
 
-  abstract createServer (
+  createServer (
     app: Express,
     config: Cfg,
     onWarning: unknown,
-  ): Bluebird<[number, WarningErr?]>
+  ): Bluebird<[number, WarningErr?]> {
+    return new Bluebird((resolve, reject) => {
+      const { port, fileServerFolder, socketIoRoute, baseUrl, experimentalSkipDomainInjection } = config
+
+      this._server = this._createHttpServer(app)
+
+      this.skipDomainInjectionForDomains = experimentalSkipDomainInjection
+
+      const onError = (err) => {
+        // if the server bombs before starting
+        // and the err no is EADDRINUSE
+        // then we know to display the custom err message
+        if (err.code === 'EADDRINUSE') {
+          return reject(this.portInUseErr(port))
+        }
+      }
+
+      debug('createServer connecting to server')
+
+      this.server.on('connect', this.onConnect.bind(this))
+      this.server.on('upgrade', (req, socket, head) => this.onUpgrade(req, socket, head, socketIoRoute))
+      this.server.once('error', onError)
+
+      this._graphqlWS = graphqlWS(this.server, `${socketIoRoute}-graphql`)
+
+      return this._listen(port, (err) => {
+        // if the server bombs before starting
+        // and the err no is EADDRINUSE
+        // then we know to display the custom err message
+        if (err.code === 'EADDRINUSE') {
+          return reject(this.portInUseErr(port))
+        }
+      })
+      .then((port) => {
+        return Bluebird.all([
+          httpsProxy.create(appData.path('proxy'), port, {
+            onRequest: this.callListeners.bind(this),
+            onUpgrade: this.onSniUpgrade.bind(this),
+          }),
+
+          fileServer.create(fileServerFolder),
+        ])
+        .spread((httpsProxy, fileServer) => {
+          this._httpsProxy = httpsProxy
+          this._fileServer = fileServer
+
+          // if we have a baseUrl let's go ahead
+          // and make sure the server is connectable!
+          if (baseUrl) {
+            this._baseUrl = baseUrl
+
+            if (config.isTextTerminal) {
+              return this._retryBaseUrlCheck(baseUrl, onWarning)
+              .return(null)
+              .catch((e) => {
+                debug(e)
+
+                return reject(errors.get('CANNOT_CONNECT_BASE_URL'))
+              })
+            }
+
+            return ensureUrl.isListening(baseUrl)
+            .return(null)
+            .catch((err) => {
+              debug('ensuring baseUrl (%s) errored: %o', baseUrl, err)
+
+              return errors.get('CANNOT_CONNECT_BASE_URL_WARNING', baseUrl)
+            })
+          }
+        }).then((warning) => {
+          // once we open set the domain to root by default
+          // which prevents a situation where navigating
+          // to http sites redirects to /__/ cypress
+          this._remoteStates.set(baseUrl != null ? baseUrl : '<root>')
```

Comment:
This is different behavior than exists in `server-ct` today - should we default to '<root>' regardless of testing type?

Reply:
Right I see, in `server-ct` we always set to `baseUrl`. I also noticed we have https://github.com/cypress-io/cypress/blob/9bd124e3089abeef818cd738fad46d10364eef12/packages/server/lib/server-base.ts#L493-L495

```js
    const baseUrl = this._baseUrl ?? '<root>'

    return this._remoteStates.set(baseUrl)
```

I think it makes no difference, ultimately we end up setting it to either `baseUrl` (always defined in CT) or `<root>`. If, for whatever reason `baseUrl` was not defined in CT, the `reset()` code I linked makes it `<root>` - so I think there's no actual change in behavior here.